src/server/engine_shard.cc

```diff
@@ -391,28 +391,53 @@ void EngineShard::Shutdown() {
 
   queue_.Shutdown();
   queue2_.Shutdown();
-  DCHECK(!fiber_periodic_.IsJoinable());
+  DCHECK(!fiber_heartbeat_periodic_.IsJoinable());
+  DCHECK(!fiber_shard_handler_periodic_.IsJoinable());
 
   ProactorBase::me()->RemoveOnIdleTask(defrag_task_);
 }
 
 void EngineShard::StopPeriodicFiber() {
-  fiber_periodic_done_.Notify();
-  if (fiber_periodic_.IsJoinable()) {
-    fiber_periodic_.Join();
+  fiber_heartbeat_periodic_done_.Notify();
+  if (fiber_heartbeat_periodic_.IsJoinable()) {
+    fiber_heartbeat_periodic_.Join();
+  }
+  fiber_shard_handler_periodic_done_.Notify();
+  if (fiber_shard_handler_periodic_.IsJoinable()) {
+    fiber_shard_handler_periodic_.Join();
   }
 }
 
-void EngineShard::StartPeriodicFiber(util::ProactorBase* pb, std::function<void()> global_handler) {
+void EngineShard::StartPeriodicFiberImpl(util::ProactorBase* pb,
+                                         std::function<void()> shard_handler, bool heartbeat) {
   uint32_t clock_cycle_ms = 1000 / std::max<uint32_t>(1, GetFlag(FLAGS_hz));
   if (clock_cycle_ms == 0)
     clock_cycle_ms = 1;
 
-  fiber_periodic_ = MakeFiber([this, index = pb->GetPoolIndex(), period_ms = clock_cycle_ms,
-                               handler = std::move(global_handler)] {
-    ThisFiber::SetName(absl::StrCat("shard_periodic", index));
-    RunPeriodic(std::chrono::milliseconds(period_ms), std::move(handler));
-  });
+  if (heartbeat) {
+    fiber_heartbeat_periodic_ =
+        MakeFiber([this, index = pb->GetPoolIndex(), period_ms = clock_cycle_ms,
+                   handler = std::move(shard_handler)]() mutable {
+          ThisFiber::SetName(absl::StrCat("heartbeat_periodic", index));
+          RunHeartbeatPeriodic(std::chrono::milliseconds(period_ms), std::move(handler));
+        });
+  } else {
+    fiber_shard_handler_periodic_ =
+        MakeFiber([this, index = pb->GetPoolIndex(), period_ms = clock_cycle_ms,
```

Comment:
see the condition in origin code RunPeriodic to run the shard handler
if (shard_handler && last_handler_ms + 100 < last_heartbeat_ms)

we run shard_handler every 100 ms this is unrelated to to FLAGS_hz
I believe this function StartPeriodicFiberImpl is rednudant

Reply:
It depends on the condition implicitly when the period is greater than 100 ms because we wait `period_ms`. So if let's say, `period_ms=500` the shard handler will run every `500ms` (we always wait period_ms between runs) so that `100` is really a minimum. I refactored this a little bit to mirror this behaviour